File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root){
      if(!root) return true;
      int r = h(root->right);
      int l = h(root->left);
      if(abs(r-l) > 1) return false;
      return isBalanced(root->left) && isBalanced(root->right);
    }
private:
    int h(TreeNode * r){
        if(!r) return 0;
        return 1 + max(h(r->left), h(r->right)); 
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    T = 0
    def dfs(root) -> int:
        if not root:
            return 0
        l = self.dfs(root.left)
        r = self.dfs(root.right)
        T = max(T,r+l)
        return 1+max(r,l)
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.dfs(root)
        return T
        